spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala

```diff
@@ -330,1243 +333,1263 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde {
+          if (leftExpr.isDefined && rightExpr.isDefined) {
+            val builder = ExprOuterClass.Or.newBuilder()
+            builder.setLeft(leftExpr.get)
+            builder.setRight(rightExpr.get)
 
-      case InitCap(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("initcap", childExpr)
+            Some(
+              ExprOuterClass.Expr
+                .newBuilder()
+                .setOr(builder)
+                .build())
+          } else {
+            None
+          }
 
-      case Length(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("length", childExpr)
+        case UnaryExpression(child) if expr.prettyName == "promote_precision" =>
+          // `UnaryExpression` includes `PromotePrecision` for Spark 3.2 & 3.3
+          // `PromotePrecision` is just a wrapper, don't need to serialize it.
+          exprToProtoInternal(child, inputs)
 
-      case Lower(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("lower", childExpr)
+        case CheckOverflow(child, dt, nullOnOverflow) =>
+          val childExpr = exprToProtoInternal(child, inputs)
 
-      case Md5(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("md5", childExpr)
+          if (childExpr.isDefined) {
+            val builder = ExprOuterClass.CheckOverflow.newBuilder()
+            builder.setChild(childExpr.get)
+            builder.setFailOnError(!nullOnOverflow)
 
-      case OctetLength(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("octet_length", childExpr)
+            // `dataType` must be decimal type
+            val dataType = serializeDataType(dt)
+            builder.setDatatype(dataType.get)
 
-      case Reverse(child) =>
-        val childExpr = exprToProtoInternal(Cast(child, StringType), inputs)
-        scalarExprToProto("reverse", childExpr)
+            Some(
+              ExprOuterClass.Expr
+                .newBuilder()
+                .setCheckOverflow(builder)
+                .build())
+          } else {
+            None
+          }
 
-      case StringInstr(str, substr) =>
-        val leftExpr = exprToProtoInternal(Cast(str, StringType), inputs)
-        val rightExpr = exprToProtoInternal(Cast(substr, StringType), inputs)
-        scalarExprToProto("strpos", leftExpr, rightExpr)
+        case attr: AttributeReference =>
+          val dataType = serializeDataType(attr.dataType)
 
-      case StringRepeat(str, times) =>
-        val leftExpr = exprToProtoInternal(Cast(str, StringType), inputs)
-        val rightExpr = exprToProtoInternal(Cast(times, LongType), inputs)
-        scalarExprToProto("repeat", leftExpr, rightExpr)
+          if (dataType.isDefined) {
```

Comment:
Hmm is the `binding` change accidentally removed?

Reply:
Oh, yea.